src/core/dyad_core.c

```diff
@@ -0,0 +1,788 @@
+    if (cons_managed_path == NULL)
+    {
+        (*ctx)->cons_managed_path = NULL;
+    }
+    else
+    {
+        (*ctx)->cons_managed_path = (char*) malloc(strlen(cons_managed_path)+1);
+        if ((*ctx)->cons_managed_path == NULL)
+        {
+            fprintf (stderr, "Could not allocate buffer for Consumer managed path!\n");
+            free((*ctx)->kvs_namespace);
+            free((*ctx)->prod_managed_path);
+            free(*ctx);
+            *ctx = NULL;
+            return DYAD_NOCTX;
+        }
+        strncpy(
+                (*ctx)->cons_managed_path,
+                cons_managed_path,
+                strlen(cons_managed_path)+1
+            );
+    }
+    // Initialization is now complete!
+    // Set reenter and initialized to indicate this.
+    printf("Initialization complete\n");
+    (*ctx)->reenter = true;
+    (*ctx)->initialized = true;
+    // TODO Print logging info
+    return DYAD_OK;
+}
+
+int dyad_produce(dyad_ctx_t *ctx, const char *fname)
+{
+    // Make sure the dyad_ctx_t object is valid for
+    // the upcoming producer operation.
+    // If not, forward the return code from validate_context
+    int rc = 0;
+    rc = validate_context(ctx, false, true);
+    if (DYAD_IS_ERROR(rc))
+    {
+        return rc;
+    }
+    // If the context is valid, call dyad_commit to perform
+    // the producer operation
+    return dyad_commit(ctx, fname);
+}
+
+int dyad_consume(dyad_ctx_t *ctx, const char *fname)
+{
+    int rc = 0;
+    // Make sure that the correct information is in ctx
+    // to run dyad_consume.
+    rc = validate_context(ctx, true, false);
+    if (DYAD_IS_ERROR(rc))
+    {
+        return rc;
+    }
+    // Set reenter to false to avoid recursively performing
+    // DYAD operations
+    ctx->reenter = false;
+    // Call dyad_fetch to get (and possibly wait on)
+    // data from the Flux KVS
+    dyad_kvs_response_t *resp = NULL;
+    rc = dyad_fetch(ctx, fname, &resp);
+    // If an error occured in dyad_fetch, log an error
+    // and return the corresponding DYAD return code
+    if (DYAD_IS_ERROR(rc))
+    {
+        DYAD_LOG_ERR (ctx, "dyad_fetch failed!\n");
+        // No need to free 'resp' because it will always
+        // be NULL if dyad_fetch doesn't return DYAD_OK
+        return rc;
+    }
+    // If dyad_fetch was successful, but resp is still NULL,
+    // then we need to skip data transfer.
+    // This will most likely happend because shared_storage
+    // is enabled
+    if (resp == NULL)
+    {
+        DYAD_LOG_INFO (ctx, "The KVS response is NULL! Skipping dyad_pull!\n");
+        return DYAD_OK;
+    }
+    // Call dyad_pull to fetch the data from the producer's
+    // Flux broker
+    rc = dyad_pull(ctx, fname, resp);
+    // Set reenter to true to allow additional intercepting
+    ctx->reenter = true;
+    // Regardless if there was an error in dyad_pull,
+    // free the KVS response object
+    dyad_free_kvs_response(&resp);
```

Comment:
I think this function can just be a part of `dyad_pull()`

Reply:
Resolved